### PR TITLE
Implement support for round tripping SSZData by storing tree nodes

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
Implements the ability to iterate an SSZData object storing information about its tree nodes and then rebuild the same object from that stored data.

"Compression" of branch nodes is supported by skipping over a given number of branch levels rather than having to store each individual branch node.

Additionally, nodes which have already be stored can be skipped from the traversal entirely. Empty nodes, padding out the end of objects or unused by lists which are not completely full are also skipped when storing and loading.

Note that this hasn't been hooked into the actual database implementation and is currently unused in production code, but it's a significant enough change on its own.

## Fixed Issue(s)
#4173 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: ssz/src/main/java/tech/pegasys/teku/ssz/schema/impl/AbstractSszPrimitiveSchema.java [hunk 1]
```diff
@@ -60,6 +63,26 @@ public SszDataT createFromBackingNode(TreeNode node) {
...
+      final int maxBranchLevelsSkipped,
+      final long rootGIndex,
+      final TreeNode node) {
+    nodeStore.storeLeafNode((LeafDataNode) node, rootGIndex);
+  }
+
+  @Override
+  public TreeNode loadBackingNodes(
+      final TreeNodeSource nodeSource, final Bytes32 rootHash, final long rootGIndex) {
+    final Bytes data = nodeSource.loadLeafNode(rootHash, rootGIndex);
```
Comment (Nashatyrev) [inline ssz/src/main/java/tech/pegasys/teku/ssz/schema/impl/AbstractSszPrimitiveSchema.java]: M.b. move `data` calc under `else` block?

**Review by Nashatyrev — APPROVED** [review]
LGTM 👍 
Didn't deep dive to implementation details, but the overall design looks cool!

File: ssz/src/main/java/tech/pegasys/teku/ssz/schema/impl/AbstractSszPrimitiveSchema.java (hunk 1 above)
Comment (ajsutton) [reply ssz/src/main/java/tech/pegasys/teku/ssz/schema/impl/AbstractSszPrimitiveSchema.java]: oh yeah, that's a nice win.

File: ssz/src/main/java/tech/pegasys/teku/ssz/schema/impl/AbstractSszListSchema.java [hunk 2]
```diff
@@ -127,6 +134,116 @@ public TreeNode sszDeserializeTree(SszReader reader) {
...
+  public void storeBackingNodes(
+      final TreeNodeStore nodeStore,
+      final int maxBranchLevelsSkipped,
+      final long rootGIndex,
+      final TreeNode node) {
+    // Lists start with a branch with the data on the left and length on the right
+    final TreeNode vectorNode = getVectorNode(node);
+    final TreeNode lengthNode = node.get(GIndexUtil.RIGHT_CHILD_G_INDEX);
+
+    // Stroe vector data (omitting empty list items at the end...)
```
Comment (rolfyone) [inline ssz/src/main/java/tech/pegasys/teku/ssz/schema/impl/AbstractSszListSchema.java]: typo - 'store'

**Review by rolfyone — APPROVED** [review]
LGTM

**Pull request merged**